stl/inc/utility

```diff
@@ -644,6 +645,82 @@ template <size_t _Idx>
 inline constexpr in_place_index_t<_Idx> in_place_index{};
 #endif // _HAS_CXX17
 
+#if _HAS_CXX20
+
+template <class _Ty>
+inline constexpr bool _Is_integral_not_bool_char = is_integral_v<_Ty>
+    && !_Is_any_of_v<_Ty, bool, char,
+#ifdef _NATIVE_WCHAR_T_DEFINED
+        wchar_t,
+#endif // _NATIVE_WCHAR_T_DEFINED
+#ifdef __cpp_char8_t
+        char8_t,
+#endif // __cpp_char8_t
+        char16_t, char32_t>;
+
+// SAFE INTEGRAL COMPARISONS
+template <class _Ty1, class _Ty2>
+constexpr bool in_range(const _Ty2 _Value) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>);
+    if constexpr (is_signed_v<_Ty1> && is_unsigned_v<_Ty2>) {
+        return _Value <= make_unsigned_t<_Ty1>((numeric_limits<_Ty1>::max)());
+    } else if constexpr (is_unsigned_v<_Ty1> && is_signed_v<_Ty2>) {
+        return _Value >= 0 && make_unsigned_t<_Ty2>(_Value) <= (numeric_limits<_Ty1>::max)();
+    } else {
+        return (numeric_limits<_Ty1>::min)() <= _Value && _Value <= (numeric_limits<_Ty1>::max)();
+    }
+}
+
+template <class _Ty1, class _Ty2>
+constexpr bool cmp_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>);
+    if constexpr (is_signed_v<_Ty1> && is_unsigned_v<_Ty2>) {
+        return _Left >= 0 && static_cast<make_unsigned_t<_Ty1>>(_Left) == _Right;

```

Comment:
I think it would make sense to reorder the clauses in all of these `meow >= 0 && convert_to_unsigned(meow) == woof` expressions as `convert_to_unsigned(meow) == woof && meow >= 0`. The latter form rejects more `false` cases - those in which `meow` is positive but differs from `woof` - earlier.